File: DevExpress.ProductsDemo.Wpf/CS/DevExpress.ProductsDemo.Wpf/MainWindowViewModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Windows;
using System.Windows.Data;
using System.Windows.Media.Imaging;
using DevExpress.Images;
using DevExpress.Mvvm;
using DevExpress.Mvvm.Native;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.UI;
using DevExpress.Utils;
using DevExpress.Xpf.Bars;
using DevExpress.Xpf.Core;
using ProductsDemo.Modules;
using System.ComponentModel.DataAnnotations;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.DevAV;

namespace ProductsDemo {
    public class MainWindowViewModel {

        public MainWindowViewModel() {
            SplashScreenType = typeof(SplashScreenWindow);
            Modules = new List<ModuleInfo>() {
                ViewModelSource.Create(() => new ModuleInfo("GridTasksModule", this, "Grid: Tasks")).SetIcon("GridTasks"),
                ViewModelSource.Create(() => new ModuleInfo("GridContactsModule", this, "Grid: Contacts")).SetIcon("GridContacts"),
                ViewModelSource.Create(() => new ModuleInfo("SpreadsheetModule", this, "Spreadsheet")).SetIcon("Spreadsheet"),
                ViewModelSource.Create(() => new ModuleInfo("RichEditModule", this, "Word Processing")).SetIcon("WordProcessing"),
                ViewModelSource.Create(() => new ModuleInfo("ReportsModule", this, "Banded Reports")).SetIcon("BandedReports"),
                ViewModelSource.Create(() => new ModuleInfo("PivotGridModule", this, "Pivot Table")).SetIcon("Pivot"),
                ViewModelSource.Create(() => new ModuleInfo("SalesDashboard", this, "Analytics")).SetIcon("Analytics"),
                ViewModelSource.Create(() => new ModuleInfo("PhotoGallery", this, "Photo Gallery Map")).SetIcon("WeatherMap"),
                ViewModelSource.Create(() => new ModuleInfo("SchedulerModule", this, "Scheduler")).SetIcon("Scheduler"),
                ViewModelSource.Create(() => new ModuleInfo("PdfViewerDemo", this, "Pdf Viewer")).SetIcon("PDFViewer")
            };
            MetadataLocator.Default = MetadataLocator.Create().AddMetadata<PrefixEnumWithExternalMetadata>();
        }
        public virtual IEnumerable<ModuleInfo> Modules { get; protected set; }

        public virtual ModuleInfo SelectedModuleInfo { get; set; }
        public virtual Type SplashScreenType { get; set; }
        public virtual int DefaultBackstatgeIndex { get; set; }
        public virtual bool HasPrinting { get; set; }
        public virtual bool IsBackstageOpen { get; set; }
        public void Exit() {
            CurrentWindowService.Close();
        }
        public void OnModulesLoaded() {
            if(SelectedModuleInfo == null) {
                SelectedModuleInfo = Modules.First();
                SelectedModuleInfo.IsSelected = true;
                SelectedModuleInfo.Show();
            }
            SplashScreenType = typeof(ProgressWindow);
            ApplicationJumpListService.Items.AddOrReplace("New Task", NewTaskIcon, ShowGridTasksModuleNewItemWindow);
            ApplicationJumpListService.Items.AddOrReplace("New Contact", NewContactIcon, ShowGridContactsModuleNewItemWindow);
            ApplicationJumpListService.Apply();
        }
        [Required]
        protected virtual ICurrentWindowService CurrentWindowService { get { return null; } }
        [Required]
        protected virtual IApplicationJumpListService ApplicationJumpListService { get { return null; } }
        [Required]
        protected virtual INavigationService NavigationService { get { return null; } }
        protected virtual void OnSelectedModuleInfoChanged() {
            if(!allowSelectedModuleInfoChanged)
                return;
            PrintingService.PrintableControlLink = null;
            SelectedModuleInfo.IsSelected = true;
            SelectedModuleInfo.Show();
        }
        protected virtual void OnIsBackstageOpenChanged() {
            HasPrinting = PrintingService.HasPrinting;
            if(!HasPrinting && DefaultBackstatgeIndex == 1)
                DefaultBackstatgeIndex = 0;
        }
        BitmapImage NewTaskIcon {
            get { return new BitmapImage(AssemblyHelper.GetResourceUri(typeof(DXImages).Assembly, "Images/Tasks/NewTask_16x16.png")); }
        }
        BitmapImage NewContactIcon {
            get { return new BitmapImage(AssemblyHelper.GetResourceUri(typeof(DXImages).Assembly, "Images/Mail/NewContact_16x16.png")); }
        }
        bool allowSelectedModuleInfoChanged = true;
        void ShowGridModuleNewItemWindow<T>(string moduleType) where T : class {
            if(Application.Current.Windows.Count != 1) return;
            GridViewModelBase<T> viewModel = ViewHelper.GetViewModelFromView(NavigationService.Current) as GridViewModelBase<T>;
            if(viewModel != null)
                viewModel.ShowNewItemWindow();
            else {
                var moduleInfo = Modules.Single(m => m.Type == moduleType);
                allowSelectedModuleInfoChanged = false;
                SelectedModuleInfo = moduleInfo;
                allowSelectedModuleInfoChanged = true;
                moduleInfo.Show(GridModuleNavigationParameter.NewItem);
            }
        }
        void ShowGridTasksModuleNewItemWindow() {
            ShowGridModuleNewItemWindow<Task>("GridTasksModule");
        }
        void ShowGridContactsModuleNewItemWindow() {
            ShowGridModuleNewItemWindow<Contact>("GridContactsModule");
        }
    }
    public class ModuleInfo {
        ISupportServices parent;

        public ModuleInfo(string _type, object parent, string _title) {
            Type = _type;
            this.parent = (ISupportServices)parent;
            Title = _title;
        }
        public string Type { get; private set; }
        public virtual bool IsSelected { get; set; }
        public string Title { get; private set; }
        public virtual Uri Icon { get; set; }
        public ModuleInfo SetIcon(string icon) {
            this.Icon = AssemblyHelper.GetResourceUri(typeof(ModuleInfo).Assembly, string.Format("Images/{0}.png", icon));
            return this;
        }
        public void Show(object parameter = null) {
            INavigationService navigationService = parent.ServiceContainer.GetRequiredService<INavigationService>();
            navigationService.Navigate(Type, parameter, parent);
        }
    }
    public class PrefixEnumWithExternalMetadata {
        public static void BuildMetadata(EnumMetadataBuilder<PersonPrefix> builder) {
            builder
                .Member(PersonPrefix.Dr)
                    .ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Doctor.png")
                .EndMember()
                .Member(PersonPrefix.Mr)
                    .ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Mr.png")
                .EndMember()
                .Member(PersonPrefix.Ms)
                    .ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Ms.png")
                .EndMember()
                .Member(PersonPrefix.Miss)
                    .ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Miss.png")
                .EndMember()
                .Member(PersonPrefix.Mrs)
                    .ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Mrs.png")
                .EndMember();
        }
    }
}

File: DevExpress.ProductsDemo.Wpf/VB/DevExpress.ProductsDemo.Wpf/MainWindowViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Windows
Imports System.Windows.Data
Imports System.Windows.Media.Imaging
Imports DevExpress.Images
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.Native
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Mvvm.UI
Imports DevExpress.Utils
Imports DevExpress.Xpf.Bars
Imports DevExpress.Xpf.Core
Imports ProductsDemo.Modules
Imports System.ComponentModel.DataAnnotations
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.DevAV

Namespace ProductsDemo
    Public Class MainWindowViewModel

        Public Sub New()
            SplashScreenType = GetType(SplashScreenWindow)
            Modules = New List(Of ModuleInfo)() From {ViewModelSource.Create(Function() New ModuleInfo("GridTasksModule", Me, "Grid: Tasks")).SetIcon("GridTasks"), ViewModelSource.Create(Function() New ModuleInfo("GridContactsModule", Me, "Grid: Contacts")).SetIcon("GridContacts"), ViewModelSource.Create(Function() New ModuleInfo("SpreadsheetModule", Me, "Spreadsheet")).SetIcon("Spreadsheet"), ViewModelSource.Create(Function() New ModuleInfo("RichEditModule", Me, "Word Processing")).SetIcon("WordProcessing"), ViewModelSource.Create(Function() New ModuleInfo("ReportsModule", Me, "Banded Reports")).SetIcon("BandedReports"), ViewModelSource.Create(Function() New ModuleInfo("PivotGridModule", Me, "Pivot Table")).SetIcon("Pivot"), ViewModelSource.Create(Function() New ModuleInfo("SalesDashboard", Me, "Analytics")).SetIcon("Analytics"), ViewModelSource.Create(Function() New ModuleInfo("PhotoGallery", Me, "Photo Gallery Map")).SetIcon("WeatherMap"), ViewModelSource.Create(Function() New ModuleInfo("SchedulerModule", Me, "Scheduler")).SetIcon("Scheduler"), ViewModelSource.Create(Function() New ModuleInfo("PdfViewerDemo", Me, "Pdf Viewer")).SetIcon("PDFViewer")}
            MetadataLocator.Default = MetadataLocator.Create().AddMetadata(Of PrefixEnumWithExternalMetadata)()
        End Sub
        Public Overridable Property Modules() As IEnumerable(Of ModuleInfo)

        Public Overridable Property SelectedModuleInfo() As ModuleInfo
        Public Overridable Property SplashScreenType() As Type
        Public Overridable Property DefaultBackstatgeIndex() As Integer
        Public Overridable Property HasPrinting() As Boolean
        Public Overridable Property IsBackstageOpen() As Boolean
        Public Sub [Exit]()
            CurrentWindowService.Close()
        End Sub
        Public Sub OnModulesLoaded()
            If SelectedModuleInfo Is Nothing Then
                SelectedModuleInfo = Modules.First()
                SelectedModuleInfo.IsSelected = True
                SelectedModuleInfo.Show()
            End If
            SplashScreenType = GetType(ProgressWindow)
            ApplicationJumpListService.Items.AddOrReplace("New Task", NewTaskIcon, AddressOf ShowGridTasksModuleNewItemWindow)
            ApplicationJumpListService.Items.AddOrReplace("New Contact", NewContactIcon, AddressOf ShowGridContactsModuleNewItemWindow)
            ApplicationJumpListService.Apply()
        End Sub
        <Required>
        Protected Overridable ReadOnly Property CurrentWindowService() As ICurrentWindowService
            Get
                Return Nothing
            End Get
        End Property
        <Required>
        Protected Overridable ReadOnly Property ApplicationJumpListService() As IApplicationJumpListService
            Get
                Return Nothing
            End Get
        End Property
        <Required>
        Protected Overridable ReadOnly Property NavigationService() As INavigationService
            Get
                Return Nothing
            End Get
        End Property
        Protected Overridable Sub OnSelectedModuleInfoChanged()
            If Not allowSelectedModuleInfoChanged Then
                Return
            End If
            PrintingService.PrintableControlLink = Nothing
            SelectedModuleInfo.IsSelected = True
            SelectedModuleInfo.Show()
        End Sub
        Protected Overridable Sub OnIsBackstageOpenChanged()
            HasPrinting = PrintingService.HasPrinting
            If (Not HasPrinting) AndAlso DefaultBackstatgeIndex = 1 Then
                DefaultBackstatgeIndex = 0
            End If
        End Sub
        Private ReadOnly Property NewTaskIcon() As BitmapImage
            Get
                Return New BitmapImage(AssemblyHelper.GetResourceUri(GetType(DXImages).Assembly, "Images/Tasks/NewTask_16x16.png"))
            End Get
        End Property
        Private ReadOnly Property NewContactIcon() As BitmapImage
            Get
                Return New BitmapImage(AssemblyHelper.GetResourceUri(GetType(DXImages).Assembly, "Images/Mail/NewContact_16x16.png"))
            End Get
        End Property
        Private allowSelectedModuleInfoChanged As Boolean = True
        Private Sub ShowGridModuleNewItemWindow(Of T As Class)(ByVal moduleType As String)
            If Application.Current.Windows.Count <> 1 Then
                Return
            End If
            Dim viewModel As GridViewModelBase(Of T) = TryCast(ViewHelper.GetViewModelFromView(NavigationService.Current), GridViewModelBase(Of T))
            If viewModel IsNot Nothing Then
                viewModel.ShowNewItemWindow()
            Else
                Dim moduleInfo = Modules.Single(Function(m) m.Type = moduleType)
                allowSelectedModuleInfoChanged = False
                SelectedModuleInfo = moduleInfo
                allowSelectedModuleInfoChanged = True
                moduleInfo.Show(GridModuleNavigationParameter.NewItem)
            End If
        End Sub
        Private Sub ShowGridTasksModuleNewItemWindow()
            ShowGridModuleNewItemWindow(Of Task)("GridTasksModule")
        End Sub
        Private Sub ShowGridContactsModuleNewItemWindow()
            ShowGridModuleNewItemWindow(Of Contact)("GridContactsModule")
        End Sub
    End Class
    Public Class ModuleInfo
        Private parent As ISupportServices

        Public Sub New(ByVal _type As String, ByVal parent As Object, ByVal _title As String)
            Type = _type
            Me.parent = DirectCast(parent, ISupportServices)
            Title = _title
        End Sub
        Private privateType As String
        Public Property Type() As String
            Get
                Return privateType
            End Get
            Private Set(ByVal value As String)
                privateType = value
            End Set
        End Property
        Public Overridable Property IsSelected() As Boolean
        Private privateTitle As String
        Public Property Title() As String
            Get
                Return privateTitle
            End Get
            Private Set(ByVal value As String)
                privateTitle = value
            End Set
        End Property
        Public Overridable Property Icon() As Uri
        Public Function SetIcon(ByVal icon As String) As ModuleInfo
            Me.Icon = AssemblyHelper.GetResourceUri(GetType(ModuleInfo).Assembly, String.Format("Images/{0}.png", icon))
            Return Me
        End Function
        Public Sub Show(Optional ByVal parameter As Object = Nothing)
            Dim navigationService As INavigationService = parent.ServiceContainer.GetRequiredService(Of INavigationService)()
            navigationService.Navigate(Type, parameter, parent)
        End Sub
    End Class
    Public Class PrefixEnumWithExternalMetadata
        Public Shared Sub BuildMetadata(ByVal builder As EnumMetadataBuilder(Of PersonPrefix))
            builder.Member(PersonPrefix.Dr).ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Doctor.png").EndMember().Member(PersonPrefix.Mr).ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Mr.png").EndMember().Member(PersonPrefix.Ms).ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Ms.png").EndMember().Member(PersonPrefix.Miss).ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Miss.png").EndMember().Member(PersonPrefix.Mrs).ImageUri("pack://application:,,,/DevExpress.ProductsDemo.Wpf;component/Images/Tasks/Mrs.png").EndMember()
        End Sub
    End Class
End Namespace
